### PR TITLE
fix MGDCTRS-800, MGDCTRS-759

### PR DESCRIPTION
**Pull request by gashcrumb**
This change overhauls the way JSON schema driven forms are rendered,
leaning more on the functionality available from the custom JSON schema
bridge.  This change also adds a simple $ref lookup function, a simple
object from schema creator and a couple quick tests for each.  Also this
change refactors the clearSecretEmptyValue function, it now returns a
new object with the changed values, and adds a test for that as well.

Hey guys, sorry this turned out to be such a big change!  But I'm hoping this starts to give us some better tools to make the JSON schema based stuff easier to work with.  Let me know if there's any issues, it could do with some extra testing!

Oh, and especially look for console.log statements, I always miss them :laughing:

### CODE REVIEW

File: src/utils/shared.ts [hunk 1]
```diff
@@ -8,10 +8,19 @@ export const mapToObject = (inputMap: Map<string, unknown>): object => {
...
-    if (_.isEmpty((obj as { [key: string]: any })[key])) {
-      (obj as { [key: string]: any })[key] = '';
+/**
+ * Returns a new object that does not contain empty objects as values
+ * and leaves the passed in value untouched.
+ * @param obj
+ * @returns
+ */
+export const clearEmptyObjectValues = (obj: any): any => {
+  const answer: any = { ...obj };
```
Comment (indraraj) [inline src/utils/shared.ts:18]: this may only do a shallow copy

File: src/app/pages/ConnectorDetailsPage/ConfigurationStep.tsx [hunk 2]
```diff
@@ -31,8 +31,9 @@ export const ConfigurationStep: FC<ConfigurationStepProps> = ({
 }) => {
   const { t } = useTranslation();
 
-  const formConfiguration = JSON.parse(JSON.stringify(configuration));
-  clearSecretEmptyValue(formConfiguration);
+  const formConfiguration = clearEmptyObjectValues(
+    JSON.parse(JSON.stringify(configuration))
```
Comment (indraraj) [inline src/app/pages/ConnectorDetailsPage/ConfigurationStep.tsx:35]: it will be better to move this to `clearEmptyObjectValues` function itself

**Comment by indraraj** [comment]
In Duplicate and edit mode if the original connector config does not have `data_shape` we have to select its value to make the next button enabled, even when in the UI it's already shown as selected.
ezgif com-gif-maker

File: src/app/pages/ConnectorDetailsPage/ConfigurationStep.tsx (hunk 2 above)
Comment (gashcrumb) [reply src/app/pages/ConnectorDetailsPage/ConfigurationStep.tsx:35]: The `JSON.parse(JSON.stringify(configuration))` part?  I was going to ask why these have been necessary, I think these configuration objects should treated as immutables, and so these kinds of deep copies shouldn't be required.  I mean, stuff happens, but yeah, revisiting these patterns and why they're needed is worthwhile as they can hide some stuff.

**Comment by gashcrumb** [comment]
> In Duplicate and edit mode if the original connector config does not have data_shape we have to select its value to make the next button enabled, even when in the UI it's already shown as selected.

Hmm, was worried about that, I fixed 2 bugs and created a new one :-)  Thanks for raising it, I'll have a look today.

File: src/utils/shared.ts (hunk 1 above)
Comment (gashcrumb) [reply src/utils/shared.ts:18]: I think it's fine provided we're all considering these configuration objects as immutable.  It really shouldn't be necessary to modify these object's attributes in the views, and even configuring should be producing a new configuration object vs directly modifying the attributes of the existing one.

**Comment by gashcrumb** [comment]
Ok, think I have that behavior fixed, should be ready for a retest.  Some of the further form updates we've gotten requests for should be pretty straightforward to address, but I'll hold off adding anything else to this PR unless required.

File: src/app/pages/ConnectorDetailsPage/ConfigurationStep.tsx (hunk 2 above)
Comment (indraraj) [reply src/app/pages/ConnectorDetailsPage/ConfigurationStep.tsx:35]: I had some doubts about considering immutables, maybe we can connect separately for this discussion.

**Comment by indraraj** [comment]
> Ok, think I have that behavior fixed, should be ready for a retest. Some of the further form updates we've gotten requests for should be pretty straightforward to address, but I'll hold off adding anything else to this PR unless required.

Thanks, stan I have tasted it and this seems to be fixed in duplicate flow, but I can still see the same [issue](https://github.com/bf2fc6cc711aee1a0c2a/cos-ui/pull/451#issuecomment-1099012420) in EDIT flow.

**Review by indraraj — APPROVED** [review]
It looks good, for the mentioned issue maybe we can pick it up in the follow up PR

**Comment by uidoyen** [comment]
@gashcrumb great work!

**Comment by gashcrumb** [comment]
Thanks @indraraj and @uidoyen!  I'll make sure to get the edit flow working as part of https://issues.redhat.com/browse/MGDCTRS-758

**Pull request merged**